[Model Averaging] Post-localSGD optimizer

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62132
* __->__ #62131

Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

Commit:
[Model Averaging] Post-localSGD optimizer

Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/optim/post_localSGD_optimizer.py:
```diff
@@ -0,0 +1,89 @@
+        pass
+
+    # TODO
+    def load_state_dict(self, state_dict):
```
Shall we raise `NotImplementedError` for these for now?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4349,6 +4350,57 @@ def _test_DistributedDataParallel_SyncBatchNorm(
+            warmup_steps = 10
+            torch.cuda.set_device(self.rank)
+            net = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(DDP_NET).cuda(),
```
Shall we test parity for more complicated model such as torchvision.models.resnet50?

SciPioneer replied:
I am afraid this will only work if torchvision is installed.

rohan-varma replied:
yep, if you want, you can gate it with `HAS_TORCHVISION` flag.

Commit:
Update on "[Model Averaging] Post-localSGD optimizer"


Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
lgtm

Commit:
Update on "[Model Averaging] Post-localSGD optimizer"


Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

[ghstack-poisoned]

Commit:
Update on "[Model Averaging] Post-localSGD optimizer"


Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

[ghstack-poisoned]
